Wrappers/Python/cil/framework/framework.py

```diff
@@ -3351,20 +3369,180 @@ def dot(self, other, *args, **kwargs):
+        
+        Parameters
+        ----------
+        reduction_function : function 
+            The unary function to be evaluated
+        direction : string or tuple of strings
+            Specify the axis or axes to calculate 'reduction_function' along using a dimension_label. 
+            Default calculates the function over the whole array
+        out: ndarray or DataContainer, optional
+            Provide an object in which to place the result. The object must have the correct dimensions and 
+            (for DataContainers) the correct dimension_labels, but the type will be cast if necessary. See 
+            [Output type determination](https://numpy.org/doc/stable/user/basics.ufuncs.html#ufuncs-output-type) 
+            for more details.
+            Default is None
+        
+        Returns
+        -------
+        scalar or ndarray
+            The result of the unary function
+        """
+
+        if kwargs.get('axis') is not None and kwargs.get('direction') is not None:
+            raise ValueError ("Incompatible arguments: specify either direction or axis")
+
+        out = kwargs.get('out')  
+        
+        if out is None:
+            if direction is None:
+                result = reduction_function(self.as_array(), *args, **kwargs)
+            else:
+                kwargs['axis'] = self.get_dimension_axis(direction)
+                result = reduction_function(self.as_array(), *args, **kwargs)
+                if isinstance(result, numpy.ndarray):
+                    new_dimensions = numpy.array(self.dimension_labels)
+                    new_dimensions = numpy.delete(new_dimensions, kwargs['axis'])
+                    result = DataContainer(result, dimension_labels=new_dimensions)
+            return result                      
+                    
+        elif issubclass(type(out), DataContainer):
+            kwargs['out'] = None
+            if direction is None:
+                result = reduction_function(self.as_array(), *args, **kwargs)
+            else:
+                kwargs['axis'] = self.get_dimension_axis(direction)
+                result = reduction_function(self.as_array(), *args, **kwargs)
+            
+            new_dimensions = numpy.array(self.dimension_labels)
+            new_dimensions = tuple(numpy.delete(new_dimensions, kwargs['axis']))
+            if result.shape == out.shape and new_dimensions == out.dimension_labels:
+                out.fill(result.astype(out.dtype)) # convert to the type given in out
+            else:
+                raise ValueError('Data mismatch: out.shape = {} result.shape = {}, out.dimension_labels = {} result.dimension_labels = {}'.format(out.shape, result.shape, out.dimension_labels, new_dimensions))   
+
+        elif issubclass(type(out), numpy.ndarray):
+            if direction is not None:
+                kwargs['axis'] = self.get_dimension_axis(direction)
+            reduction_function(self.as_array(), *args, **kwargs)
```

Comment:
I'm not sure we care about handling this case. But it's probably fine to include!